R/transformed.R

```diff
@@ -19,14 +28,28 @@
 #' quantile(dist, 0.1) # qlnorm(0.1, 0, 0.5)
 #' generate(dist, 10) # rlnorm(10, 0, 0.5)
 #'
+#' # provide a derivative of the inverse function to avoid numerical differentiation
+#' dist <- dist_transformed(dist_normal(0, 0.5), exp, log, function(x) 1/x)
+#'
 #' @export
-dist_transformed <- function(dist, transform, inverse){
+dist_transformed <- function(dist, transform, inverse, d_inverse = NULL){
   vec_is(dist, new_dist())
-  if(is.function(transform)) transform <- list(transform)
-  if(is.function(inverse)) inverse <- list(inverse)
-  new_dist(dist = vec_data(dist),
-           transform = transform, inverse = inverse,
-           dimnames = dimnames(dist), class = "dist_transformed")
+  if (is.function(transform)) transform <- list(transform)
+  if (is.function(inverse)) inverse <- list(inverse)
+  if (is.null(d_inverse)) {
+    d_inverse <- lapply(inverse, function(inv) symbolic_derivative(inv))
+  } else if (is.function(d_inverse)) {
+    d_inverse <- list(d_inverse)
+  }
+
+  args <- vctrs::vec_recycle_common(dist = dist, transform = transform, inverse = inverse, d_inverse = d_inverse)
+  args <- transpose(args)
+  funs <- transpose(lapply(args, function(x) add_transform(x$dist, x)))
+  dist <- lapply(args, function(x) if (inherits(x$dist, "dist_transformed")) x$dist$dist else x$dist)
```

Comment:
I addressed the other issues, but I'm not sure what you mean by this. The transposition of inputs is because users can supply any of the following:

```r
dist_transformed(dist_normal(), list(exp, log, sqrt), list(log, exp, function(x) x^2))
dist_transformed(c(dist_normal(), dist_gamma(0,1)), exp, log)
dist_transformed(c(dist_normal(), dist_gamma(0,1)), list(exp, log), list(log, exp))
```

maybe you have something else in mind to deal with these different cases, but that's what I figured out

Reply:
Thanks, it's my misunderstanding! This code solves a different problem currently in the package which is why the code appears more complex.

``` r
library(distributional)
dist_transformed(dist_transformed(dist_normal(), exp, log), log, exp)
#> <distribution[1]>
#> [1] t(t(N(0, 1)))
```

<sup>Created on 2024-04-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>